api/src/main/java/org/openmrs/module/reporting/query/visit/evaluator/BasicVisitQueryEvaluator.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations
+ * under the License.
+ *
+ * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+ */
+
+package org.openmrs.module.reporting.query.visit.evaluator;
+
+import java.util.List;
+
+import org.openmrs.Visit;
+import org.openmrs.annotation.Handler;
+import org.openmrs.module.reporting.common.ListMap;
+import org.openmrs.module.reporting.common.ObjectUtil;
+import org.openmrs.module.reporting.common.TimeQualifier;
+import org.openmrs.module.reporting.evaluation.EvaluationContext;
+import org.openmrs.module.reporting.evaluation.EvaluationException;
+import org.openmrs.module.reporting.evaluation.context.VisitEvaluationContext;
+import org.openmrs.module.reporting.evaluation.querybuilder.HqlQueryBuilder;
+import org.openmrs.module.reporting.evaluation.service.EvaluationService;
+import org.openmrs.module.reporting.query.visit.VisitIdSet;
+import org.openmrs.module.reporting.query.visit.VisitQueryResult;
+import org.openmrs.module.reporting.query.visit.definition.BasicVisitQuery;
+import org.openmrs.module.reporting.query.visit.definition.VisitQuery;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Handler(supports = BasicVisitQuery.class)
+public class BasicVisitQueryEvaluator implements VisitQueryEvaluator {
+
+    @Autowired
+	EvaluationService evaluationService;
+
+    @Override
+    public VisitQueryResult evaluate(VisitQuery definition, EvaluationContext context) throws EvaluationException {
+        context = ObjectUtil.nvl(context, new EvaluationContext());
+
+        BasicVisitQuery query = (BasicVisitQuery) definition;
+        VisitQueryResult result = new VisitQueryResult(query, context);
+
+		if (context.getBaseCohort() != null && context.getBaseCohort().isEmpty()) {
+			return result;
+		}
+		if (context instanceof VisitEvaluationContext) {
+			VisitIdSet baseVisits = ((VisitEvaluationContext) context).getBaseVisits();
+			if (baseVisits != null && baseVisits.getMemberIds().isEmpty()) {
+				return result;
+			}
+		}
+
+		HqlQueryBuilder q = new HqlQueryBuilder();
+		q.select("v.visitId");
+		q.from(Visit.class, "v");
+		q.whereIn("v.visitType", query.getVisitTypes());
+		q.whereGreaterOrEqualTo("v.startDatetime", query.getStartedOnOrAfter());
+		q.whereLess("v.startDatetime", query.getStartedBefore());
+		q.whereGreaterOrEqualTo("v.stopDatetime", query.getEndedOnOrAfter());
+		q.whereLess("v.stopDatetime", query.getEndedBefore());
+		q.whereIn("v.location", query.getLocationList());
+		q.whereEncounterIn("v.visitId", context);
+
+		if (query.getWhich() == null || query.getWhich() == TimeQualifier.ANY) {
+			List<Integer> results = evaluationService.evaluateToList(q, Integer.class, context);
+			result.getMemberIds().addAll(results);
+		}
+		else {
+			q.innerJoin("v.patient", "p");
```

Comment:
It isn't 100% clear to me what this is trying to accomplish.  What does it mean for a Query to search on LAST or FIRST.  I would think that it would mean that a Visit should be included if the First or Last one for the patient matches the rest of the specified criteria.  So, I'd think we would need to limit the overall query to just look at the first or last visits for a given patient.  Is that what is happening here?  Or is there a different interpretation of this field?  If there is not an immediate requirement to support the Which/WhichNumber properties, I'd vote to hold off on them, since I think there is some ambiguity around them perhaps...